### PR TITLE
Warnings

### PR DESCRIPTION
**Pull request by GerjonM**
Adds a warning hook to fieldaccessors
Adds an isWarningFree hook to formstates

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+# 0.20.0
+
+-   Added a warning hook in the fieldAccessors, allowing you to define
```
Comment (faassen) [inline CHANGES.md]: "in field accessors"

File: CHANGES.md [hunk 2]
```diff
@@ -1,3 +1,9 @@
...
+    warning that can exist next to the erros. The form state now includes
```
Comment (faassen) [inline CHANGES.md]: "a warning"   "erros" -> "errors"

I like "that can exist in addition to the errors" better.

File: CHANGES.md [hunk 3]
```diff
@@ -1,3 +1,9 @@
...
+    an "isWarningFree" hook to deremine if the form state contains warning.
```
Comment (faassen) [inline CHANGES.md]: that's not a hook, it's a method. And 'deremine'. "contains any warning".

File: README.md [hunk 4]
```diff
@@ -686,6 +686,27 @@ such behavior for `hidden` or `required`; use `accessor.hidden` and
 ``accessor.required` in your rendering code to determine whether a field wants
 to be hidden or required.
 
+## Warnings
+
+mstform has a hook which allows you to include `warning` messages in the fields.
+Warnings are similar to errors, however, the philosophy behind it is that you
```
Comment (faassen) [inline README.md]: a "philosophy" is a bit big. "Warnings are similar to errors, but don't make the form invalid. The idea is that you can show warnings for certain fields in your form as a notification to the user". Or somethingl ike that.

File: src/state.ts [hunk 5]
```diff
@@ -81,6 +89,9 @@ export class FormState<M, D extends FormDefinition<M>> extends FormAccessorBase<
   isReadOnlyFunc: FieldAccessorAllows;
   isRequiredFunc: FieldAccessorAllows;
   isRepeatingFormDisabledFunc: RepeatingFormAccessorAllows;
+  getErrorFunc: ErrorOrWarning;
+  getWarningFunc: ErrorOrWarning;
+  isWarningFreeFunc: FieldAccessorAllows;
```
Comment (faassen) [inline src/state.ts]: This isn't a hook, it should just a method. I see you don't actually use this isWarningFreeFunc hook anywhere, and that's good at it should be removed.

File: test/warning.test.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  const o = M.create({ foo: "FOO" });
+
+  const state = form.state(o, {
+    getError: accessor => (accessor.path === "/foo" ? "Wrong" : undefined)
+  });
+  const fooField = state.field("foo");
+
+  expect(fooField.raw).toEqual("FOO");
+  expect(fooField.error).toEqual("Wrong");
+  expect(state.isWarningFree).toBeTruthy();
```
Comment (faassen) [inline test/warning.test.ts]: you can also check that the form isn't valid, and that save returns false.

File: test/warning.test.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
+import { configure } from "mobx";
```
Comment (faassen) [inline test/warning.test.ts]: One test I'm still missing is to verify that a client-side error (for instance if you set 'required' to true, or add in a custom validation function) overrules the getError hook derived error.

File: CHANGES.md [hunk 8]
```diff
@@ -1,3 +1,9 @@
...
+
```
Comment (faassen) [inline CHANGES.md]: I don't see anything about the getError hook.

File: README.md [hunk 9]
```diff
@@ -686,6 +686,27 @@ such behavior for `hidden` or `required`; use `accessor.hidden` and
 ``accessor.required` in your rendering code to determine whether a field wants
 to be hidden or required.
 
+## Warnings
```
Comment (faassen) [inline README.md]: The section about validation errors should have a mention of the get error hook. You should add a section about the getError hook. You should mention that if there is an internally generated error message that trumps the one returned by the getError hook.

File: README.md [hunk 10]
```diff
@@ -594,6 +594,20 @@ second entry, the error structure should look like this:
 }
 ```
 
+Additionally, you can assign errors to a field using your own functions.
+
+```js
+this.formState = form.state(o, {
```
Comment (faassen) [inline README.md]: Hm, this is a bit misleading, as it makes it seem like the getError hook is primarily about validation, but it shouldn't really be used for validation (we already have validation for that). I'd rather have an example that uses .path.

File: README.md [hunk 11]
```diff
@@ -594,6 +594,20 @@ second entry, the error structure should look like this:
 }
 ```
 
+Additionally, you can assign errors to a field using your own functions.
```
Comment (faassen) [inline README.md]: "using your own functions" is a bit vague. "Additionally you can also assign errors to a field that are managed outside of mstform".

File: README.md [hunk 12]
```diff
@@ -594,6 +594,20 @@ second entry, the error structure should look like this:
...
+
+```js
+this.formState = form.state(o, {
+    getError: accessor => accessor.raw === "wrong" ? "Is wrong" : undefined
+})
+```
+
+The `error` property of the field will contain the "Is wrong" error message if
+the field return `true` with the `getError` function. If a field contains both
+an internally generated error message and one that is generated via `getErrors`,
```
Comment (faassen) [inline README.md]: it's "getError", not "getErrors"

File: README.md [hunk 13]
```diff
@@ -594,6 +594,20 @@ second entry, the error structure should look like this:
...
+Additionally, you can assign errors to a field using your own functions.
+
+```js
+this.formState = form.state(o, {
+    getError: accessor => accessor.raw === "wrong" ? "Is wrong" : undefined
+})
+```
+
+The `error` property of the field will contain the "Is wrong" error message if
+the field return `true` with the `getError` function. If a field contains both
```
Comment (faassen) [inline README.md]: that's both not grammatical ("if the field return") and incorrect. The getError function (not the field) needs to return a string or undefined, not 'true'.

File: src/state.ts [hunk 14]
```diff
@@ -55,6 +59,10 @@ export interface FormStateOptions<M> {
   isRepeatingFormDisabled?: RepeatingFormAccessorAllows;
   isRequired?: FieldAccessorAllows;
 
+  getError?: ErrorOrWarning;
+  getWarning?: ErrorOrWarning;
+  isWarningFree?: FieldAccessorAllows;
```
Comment (faassen) [inline src/state.ts]: isWarningFree is not a legitimate option.

File: test/warning.test.ts (hunk 6 above)
Comment (faassen) [reply test/warning.test.ts]: We do want to check that 'validate' actually still returns true even if there are warnings.

Comment (GerjonM) [reply test/warning.test.ts]: The "a simple warning" test actually verifies this

File: README.md (hunk 10 above)
Comment (GerjonM) [reply README.md]: Used .path in the example

**Pull request merged**